walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -13,11 +13,93 @@
 // limitations under the License.
 
 package com.google.sps;
-
+import java.io.*;
+import java.util.*;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> attendees = request.getAttendees();
+    long duration = request.getDuration();
+    Collection<TimeRange> ranges = new ArrayList<>();
+
+    //--------------------------------------------------------------------edge cases
+    // if duration too long
+    if (duration > TimeRange.WHOLE_DAY.duration()) {
+      return ranges;
+    }
+
+    // if no attendees
+    if (attendees.isEmpty()) {
+      ranges.add(TimeRange.WHOLE_DAY);
+      return ranges;
+    }
+
+    //----------------------------------------------------------------- general case
+    if (events.size() >= 0) {
+      // iterate through each element in collection
+      Iterator<Event> itr = events.iterator();
+      ArrayList<ArrayList<Integer>> meetings = new ArrayList<ArrayList<Integer>>();
+      while (itr.hasNext()) {
+        // add the start time and the end time to the 2D arraylist
+        Event element = itr.next();
+        Set<String> peopleInEvent = element.getAttendees();
+        // check if any person in the event matches a person in the request
+        List<String> common = new ArrayList<String>(attendees);
+        common.retainAll(peopleInEvent);
+        // if it does then add it to meetings
+        if (common.size() > 0) {
+          TimeRange meetingDetails = element.getWhen();
+          ArrayList<Integer> temp = new ArrayList<Integer>(2);
+          temp.add(meetingDetails.start());
+          temp.add(meetingDetails.end());
+          meetings.add(temp);
+        }
```

Comment:
Rather than `ArrayList<ArrayList<Integer>>`, you could've used an `ArrayList<TimeRange>` because a `TimeRange` already includes the start time and end time (implied from the start time + duration). Plus, `getWhen()` already returns a `TimeRange`